src/goto-analyzer/build_analyzer.cpp

```diff
@@ -0,0 +1,164 @@
+    {
+      df = util_make_unique<variable_sensitivity_domain_factoryt>(
+        vs_object_factory, vsd_config);
+    }
+    // non-null is not fully supported, despite the historical options
+    // dependency-graph is quite heavily tied to the legacy-ait infrastructure
+    // dependency-graph-vs is very similar to dependency-graph
+
+    // Build the storage object
+    std::unique_ptr<ai_storage_baset> st = nullptr;
+    if(options.get_bool_option("one-domain-per-history"))
+    {
+      st = util_make_unique<history_sensitive_storaget>();
+    }
+    else if(options.get_bool_option("one-domain-per-location"))
+    {
+      st = util_make_unique<location_sensitive_storaget>();
+    }
+
+    // Only try to build the abstract interpreter if all the parts have been
+    // correctly specified and configured
+    if(hf != nullptr && df != nullptr && st != nullptr)
+    {
+      if(options.get_bool_option("recursive-interprocedural"))
+      {
+        return util_make_unique<ai_recursive_interproceduralt>(
+          std::move(hf), std::move(df), std::move(st));
+      }
+      else if(options.get_bool_option("three-way-merge"))
+      {
+        // Only works with VSD
+        if(options.get_bool_option("vsd"))
+        {
+          return util_make_unique<ai_three_way_merget>(
+            std::move(hf), std::move(df), std::move(st));
+        }
+      }
+    }
+  }
+  else if(options.get_bool_option("legacy-ait"))
+  {
+    if(options.get_bool_option("constants"))
+    {
+      // constant_propagator_ait derives from ait<constant_propagator_domaint>
+      return util_make_unique<constant_propagator_ait>(
+        goto_model.goto_functions);
+    }
+    else if(options.get_bool_option("dependence-graph"))
+    {
+      return util_make_unique<dependence_grapht>(ns);
+    }
+    else if(options.get_bool_option("dependence-graph-vs"))
+    {
+      return util_make_unique<variable_sensitivity_dependence_grapht>(
+        goto_model.goto_functions, ns, vs_object_factory, vsd_config);
+    }
+    else if(options.get_bool_option("vsd"))
+    {
+      auto df = util_make_unique<variable_sensitivity_domain_factoryt>(
+        vs_object_factory, vsd_config);
+      return util_make_unique<ait<variable_sensitivity_domaint>>(std::move(df));
+    }
+    else if(options.get_bool_option("intervals"))
+    {
+      return util_make_unique<ait<interval_domaint>>();
+    }
+#if 0
+    // Not actually implemented, despite the option...
+    else if(options.get_bool_option("non-null"))
+    {
+      return util_make_unique<ait<non_null_domaint> >();
+    }
+#endif
+  }
+  else if(options.get_bool_option("legacy-concurrent"))
+  {
+#if 0
+    // Very few domains can work with this interpreter
+    // as it requires that changes to the domain are
+    // 'non-revertable' and it has merge shared
+    if(options.get_bool_option("dependence-graph"))
+    {
+      return util_make_unique<dependence_grapht>(ns);
+    }
+#endif
```

Comment:
Not as an invariant because it is user options.  `invalid_command_line_argument_exceptiont` maybe but see above for what error handling already exists.

Reply:
Here and above: with that error handling already in place, perhaps even more of an indication that this should be `UNIMPLEMENTED` or `UNHANDLED_CASE`?

Reply:
I don't think it should be an invariant.  I think the correct thing to do is to prune the command-line options for `goto-analyzer` but that shouldn't be in a refactor commit.